bgpd/bgp_route.c

```diff
@@ -9775,32 +9775,43 @@ static int bgp_show_table(struct vty *vty, struct bgp *bgp, safi_t safi,
 	unsigned long total_count = 0;
 	struct prefix *p;
 	json_object *json_paths = NULL;
-	int first = 1;
+	json_object *json_path = NULL;
+	json_object *json_rd_desc = NULL;
+	json_object *json_rd_val = NULL;
+	bool use_json = CHECK_FLAG(show_flags, BGP_SHOW_OPT_JSON);
+	bool wide = CHECK_FLAG(show_flags, BGP_SHOW_OPT_WIDE);
+	bool all = CHECK_FLAG(show_flags, BGP_SHOW_OPT_AFI_ALL);
+	bool print_output = false;
 
 	if (output_cum && *output_cum != 0)
 		header = 0;
 
-	if (use_json && !*json_header_depth) {
-		vty_out(vty,
-			"{\n \"vrfId\": %d,\n \"vrfName\": \"%s\",\n \"tableVersion\": %" PRId64",\n \"routerId\": \"%s\",\n \"defaultLocPrf\": %u,\n"
-			" \"localAS\": %u,\n \"routes\": { ",
-			bgp->vrf_id == VRF_UNKNOWN ? -1 : (int)bgp->vrf_id,
+	if (use_json) {
+		if (!json) {
+			print_output = true;
+			json = json_object_new_object();
+		}
+		json_object_int_add(
+			json, "vrfId",
+			bgp->vrf_id == VRF_UNKNOWN ? -1 : (int)bgp->vrf_id);
+		json_object_string_add(
+			json, "vrfName",
 			bgp->inst_type == BGP_INSTANCE_TYPE_DEFAULT
-						? VRF_DEFAULT_NAME
-						: bgp->name,
-			table->version, inet_ntoa(bgp->router_id),
-			bgp->default_local_pref, bgp->as);
-		*json_header_depth = 2;
+				? VRF_DEFAULT_NAME
+				: bgp->name);
+		json_object_int_add(json, "tableVersion", table->version);
+		json_object_string_add(json, "routerId",
+				       inet_ntoa(bgp->router_id));
+		json_object_int_add(json, "defaultLocPrf",
+				    bgp->default_local_pref);
+		json_object_int_add(json, "localAS", bgp->as);
+		json_paths = json_object_new_array();
```

Comment:
You create an array here, while the original structure contains an object here. For instance:

```
 "routes": { "10.0.0.0/24": [
  {
```
vs.
```
  "routes":[
    "10.0.0.0\/24",
    [
      {
```

Reply:
Right. because in the loop if i add the array to an object, after exiting the loop its having only the last entry.

Reply:
Well, then it looks like here needs a refactor.

Reply:
@ton31337 
could you please suggest on how to proceed with this PR.
Shall I revert the changes to manual crafting and take JSON-C refactor in separate issue?

Reply:
We can proceed with old way for this feature.

Reply:
Ok. Will revert the changes back and send review request. Thanks.

Reply:
> 
> 
> We can proceed with old way for this feature.

@ton31337 
reverted the changes back to the old way. Please check